Honour joining model order in `has_many :through` associations when eager loading

yasyf opened the pull request:
Honour the order of the joining model in a `has_many :through` association when eager loading.

Previously, eager loading a `has_many :through` association with no
defined order would return the records in the natural order of the
database. Now, these records will be returned in the order that the
joining record is returned, in case there is a defined order there.

Fixes #17864.

sgrif commented on activerecord/lib/active_record/associations/preloader/association.rb:
```diff
@@ -29,7 +29,14 @@ def scope
+          if scope.values[:order].any?
+            records
+          else
+            records.sort_by { |record| ids.index record[association_key_name] }
```
Can we do this at the database level, instead? This will slow down reading through associations quite a bit. I think it would make more sense to implement this by using the order from the other scope if it's undefined on the through association.

pixeltrix replied:
I think it's going to be one of those situations where it depends on number of records in memory and size of the table since it'll mean joining the through association table to the destination table for the ordering to occur. We should also look at what happens when you have a nesting of more than one level - if we're preloading a nested through association we could end up with an expensive query.

sgrif commented:
I've added this to the 4.1 milestone, but I'm not sure that we actually want to make this change on 4.1 and 4.2, since we've released several versions with the old behavior, and this could be considered a breaking change at this point. It might make more sense to only include it in 5.0

yasyf commented:
Tried something like the following in an attempt to do the sorting at a database level, but after speaking with @matthewd it became clear that a working version of this would make the query very ugly, as the scope on the `has_many :through` can be anything at all. So I instead made the in-memory solution more efficient by pre-calculating a map of each `id` to its index.

```diff
diff --git a/activerecord/lib/active_record/associations/preloader/through_association.rb b/activerecord/lib/active_record/associations/preloader/through_association.rb
index 56aa23b..bf1713b 100644
--- a/activerecord/lib/active_record/associations/preloader/through_association.rb
+++ b/activerecord/lib/active_record/associations/preloader/through_association.rb
@@ -27,7 +27,7 @@ module ActiveRecord
 
           preloaders = preloader.preload(middle_records,
                                          source_reflection.name,
-                                         reflection_scope)
+                                         reflection_scope_with_type)
 
           @preloaded_records = preloaders.flat_map(&:preloaded_records)
 
@@ -59,6 +59,18 @@ module ActiveRecord
 
         private
 
+        def reflection_scope_with_type
+          return unless reflection_scope
+          # If dealing with a `has_many :through` association, respect the order on
+          # `reflection_scope` if it exists, else use the order of the through association
+          if !reflection_scope.values[:order] && source_reflection.belongs_to? && !through_reflection.has_one?
+            association = owners.first.association(through_reflection.name)
+            reflection_scope.joins(through_scope.table.name.to_sym).order(association.scope.values[:order])
+          else
+            reflection_scope
+          end
+        end
+
         def reset_association(owners, association_name)
           should_reset = (through_scope != through_reflection.klass.unscoped) ||
              (reflection.options[:source_type] && through_reflection.collection?)
```

eileencodes commented on activerecord/test/cases/associations/eager_test.rb:
```diff
@@ -743,6 +743,18 @@ def test_eager_with_invalid_association_reference
+    post2 = Post.create!(:title => 'Pools', :body => "I like pools!")
+
+    Tagging.create!(:taggable_type => 'Post', :taggable_id => post1.id, :tag => tag)
+    Tagging.create!(:taggable_type => 'Post', :taggable_id => post2.id, :tag => tag)
```
New tests should use the Ruby 1.9 hash syntax not hash rockets. So instead these should be changed to be like this: 

`Tagging.create!(taggable_type: 'Post', taggable_id: post2.id, tag: tag)`

eileencodes commented on activerecord/test/models/tag.rb:
```diff
@@ -5,3 +5,9 @@ class Tag < ActiveRecord::Base
+class OrderedTag < Tag
+  self.table_name = "tags"
+
+  has_many :taggings, -> { order('taggings.id DESC') }, :foreign_key => 'tag_id'
```
Here should have Ruby 1.9 syntax as well:

`has_many :taggings, -> { order('taggings.id DESC') }, foreign_key: 'tag_id'`

yasyf commented:
@eileencodes thanks, I fixed up those syntax issues.

jturkel commented:
@yasyf - I'm psyched to see you picked up this issue. Thank you! I tried out your fix but it doesn't seem to handle the case where an associated model may appear multiple times but in different orders. You can reproduce this problem with the following tweak to your test:

```
tag1 = OrderedTag.create!(name: 'Bar')
tag2 = OrderedTag.create!(name: 'Foo')

post1 = Post.create!(title: 'Beaches', body: "I like beaches!")
post2 = Post.create!(title: 'Pools', body: "I like pools!")

Tagging.create!(taggable: post1, tag: tag1)
Tagging.create!(taggable: post2, tag: tag1)
Tagging.create!(taggable: post2, tag: tag2)
Tagging.create!(taggable: post1, tag: tag2)

tags_with_includes = OrderedTag.where(id: [tag1, tag2].map(&:id)).includes(:tagged_posts).order(:id).to_a
tag1_with_includes = tags_with_includes.first
tag2_with_includes = tags_with_includes.last

assert_equal([post2, post1].map(&:title), tag1_with_includes.tagged_posts.map(&:title))
assert_equal([post1, post2].map(&:title), tag2_with_includes.tagged_posts.map(&:title))
```

yasyf commented:
@jturkel I don't know whether that is solvable using this technique, as its a completely different code path. The issue we're solving here is sorting the records as they come from the database, the one and only time they are preloaded. Once this occurs, we aren't hitting this code anymore, so the order will always be based on the first associated record that is preloaded. I can't think of how to solve this, given that when you access the `tagged_posts` of `tag2`, they have already been preloaded and ordered. Would love to hear any thoughts/suggestions you have!

jturkel commented:
I dug into commit e5299c1ef693ef434f55811027a7da975cd55ba5 that introduced #17864 and I'm pretty sure the problem is the `sort_by` at the end of `ActiveRecord::Associations::Preloader::ThroughAssociation#associated_records_by_owner` which sorts the source records based on the order they were retrieved from the database. Only doing this sort if the has_many through association has a sort (i.e. wrapping the `sort_by` in a `reflection_scope.values[:order].present?` conditional) seems to fix the problem. 

Note I'm far from an expert on the Rails codebase so definitely take my analysis with a grain of salt :)

yasyf commented:
@jturkel preventing the sort from occurring during certain conditions is definitely a better idea than simply undoing it, which is what my original solution was doing. Thanks!

jturkel commented:
@sgrif - It would be awesome to see this change make it into 4.1/4.2 since 5.0 is while away. Hopefully backwards compatibility isn't an issue since previously when no sort was specified records were returned in a non-deterministic order (i.e. whatever order the database decides to return them in which isn't guaranteed to be deterministic).

jturkel commented on activerecord/lib/active_record/associations/preloader/association.rb:
```diff
@@ -74,6 +74,12 @@ def options
 
         private
 
+        def id_to_index_map(ids)
```
Should this be pushed into `ActiveRecord::Associations::Preloader::ThroughAssociation` since it's only used there?

yasyf replied:
@jturkel not sure. I left it there because it seems like something that could viably be reused by other subclasses.

iHiD commented:
Thanks for fixing this :blue_heart: 

I'd definitely see it as a regression that warrants an earlier release than 5 (for what my opinion is worth :))

yasyf committed:
Honour the order of the joining model in a `has_many :through`
association when eager loading.

Previously, eager loading a `has_many :through` association with no
defined order would return the records in the natural order of the
database. Now, these records will be returned in the order that the
joining record is returned, in case there is a defined order there.

rafaelfranca commented:
@sgrif could you review this one? I'm removing from the milestone since I agree that it may be a breaking change.

jturkel commented:
@sgrif - Any chance this could make it into 5.0? Thanks!
